clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -418,6 +422,14 @@ public void updateLastStableOffset(TopicPartition tp, long lastStableOffset) {
         assignedState(tp).lastStableOffset = lastStableOffset;
     }
 
+    public void updatePreferredReadReplica(TopicPartition tp, int preferredReadReplicaId) {
+        assignedState(tp).preferredReadReplica = preferredReadReplicaId;
+    }
+
+    public void clearPreferredReadReplica(TopicPartition tp) {
```

Comment:
I think the main thing to figure out is in what cases we should switch back to the leader. We know about out of range errors already. Here are a few other cases:

1. Leader changes. We have to do validation anyway, so we can reset the preferred read replica.
2. The preferred read replica is marked offline in the metadata.
3. We notice changes to either the replica set or the ISR.

Possibly we could just do 1 and 2. For 3, maybe we could handle it using a periodic timer to check preferred read replica status. One option would be to couple this with `metadata.max.age.ms`, which is what the original KIP proposed.

Reply:
1 and 2 are simple enough to add. I'll do that

A timer on the client for 3 seems reasonable. So, we simply go back to the leader if it's been longer than  `metadata.max.age.ms` since our last preferred_read_replica update?

I believe we should also be able to use the broker to inform the consumer when to go back to the leader. As written, the consumer will update the preferred replica only if the preferred_read_replica field is set in the FetchResponse. This means a follower can exclude this field to remain the preferred replica. So if we want to the consumer to revert to the leader in a non-exceptional case we can set the preferred_read_replica to the leader in the response. This could be controlled by a timer or by observing ISR changes on the broker.